sycl/test-e2e/Matrix/joint_matrix_opt_kernel_feature_impl.hpp

```diff
@@ -95,8 +100,9 @@ int main() {
     init_and_multiply<bfloat16, float, 2, 1, 500,
                       16>(); // 500 is not correct size
   } catch (const sycl::exception &e) {
-    if (e.code() == errc::kernel_not_supported)
+    if (e.code() == errc::invalid)
```

Comment:
I think this test requires more investigation of why exactly error code changed. I would exclude the change from the current PR (which is already huge), and then submit the change as separate PR after investigation.
My theory is, when we use `get_sg_size`, it causes kernel to be compiled by IGC, before the check in SYCL runtime for matrix kernel feature happens, and compilation by IGC fails, because parameters are not supported, hence we are getting this `invalid` error code.
I would suggest to try doing incremental change to this test and see at which point error code changes, and then you can debug and find out the root cause. Then we can decide, if we want to change something in runtime behavior or if it is fine to report `errc::invalid` in this case...
You can also consult with author of this test, as he likely implemented the feature as well.

Reply:
As per discussion with @steffenlarsen this is an expected change of error-codes according to the spec. 
We are now catching the incompatibility earlier in the flow. When we were throwing the `errc::kernel_not_supported` we were doing so at launch time, due to
```
[Considered incompatible if: ]
It uses optional features which are not supported on the device, as described in [Section 5.7](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:optional-kernel-features)
```

Now with my change we are making a call to `get_kernel_bundle()` before we launch the kernel. This call fails with:
``` 
[Throws] An `exception` with the `errc::invalid` error code if any of the kernels identified by `kernelIds` are incompatible with all devices in `devs`.
```